File: 2016/golang/src/day1/main.go
package main

import (
	"math"
	"strconv"
	"strings"
	"utils"
)

var data = getData(utils.GetLines("input.txt"))

func contains(s [][]int, i []int) bool {
	for _, v := range s {
		if v[0] == i[0] && v[1] == i[1] {
			return true
		}
	}

	return false
}

func moveCurPos(curPos []int, curDir *int, dir string) {
	if string(dir[0]) == "L" {
		*curDir = (*curDir - 1) % 4
	} else {
		*curDir = (*curDir + 1) % 4
	}
	idx, _ := strconv.Atoi(dir[1:])
	if *curDir == 2 || *curDir == 3 || *curDir == -1 || *curDir == -2 {
		idx *= -1
	}
	curIdx := 0
	if *curDir%2 == 0 {
		curIdx = 1
	}
	curPos[curIdx] += idx
}

func getDistance(pos []int) int {
	return int(math.Abs(float64(pos[0]))) + int(math.Abs(float64(pos[1])))
}

func getData(lines []string) []string {
	return strings.Split(strings.ReplaceAll(lines[0], " ", ""), ",")
}

func partOne() string {
	curPos := make([]int, 2)
	curDir := 0 // 0=N, 1=E, 2=S, 3=W
	for _, dir := range data {
		moveCurPos(curPos, &curDir, dir)
	}
	return strconv.Itoa(getDistance(curPos))
}

func partTwo() string {
	visitedPos := make([][]int, 0)
	curPos := make([]int, 2)
	visitedPos = append(visitedPos, []int{0, 0})
	curDir := 0 // 0=N, 1=E, 2=S, 3=W
	for _, dir := range data {
		oldPos := []int{curPos[0], curPos[1]}
		moveCurPos(curPos, &curDir, dir)
		if curPos[0] == oldPos[0] {
			if curPos[1] > oldPos[1] {
				for i := oldPos[1] + 1; i <= curPos[1]; i++ {
					tmpPos := []int{curPos[0], i}
					if contains(visitedPos, tmpPos) {
						return strconv.Itoa(getDistance(tmpPos))
					}
					visitedPos = append(visitedPos, tmpPos)
				}
			} else {
				for i := oldPos[1] - 1; i >= curPos[1]; i-- {
					tmpPos := []int{curPos[0], i}
					if contains(visitedPos, tmpPos) {
						return strconv.Itoa(getDistance(tmpPos))
					}
					visitedPos = append(visitedPos, tmpPos)
				}
			}
		} else {
			if curPos[0] > oldPos[0] {
				for i := oldPos[0] + 1; i <= curPos[0]; i++ {
					tmpPos := []int{i, curPos[1]}
					if contains(visitedPos, tmpPos) {
						return strconv.Itoa(getDistance(tmpPos))
					}
					visitedPos = append(visitedPos, tmpPos)
				}
			} else {
				for i := oldPos[0] - 1; i >= curPos[0]; i-- {
					tmpPos := []int{i, curPos[1]}
					if contains(visitedPos, tmpPos) {
						return strconv.Itoa(getDistance(tmpPos))
					}
					visitedPos = append(visitedPos, tmpPos)
				}
			}
		}
	}
	return ""
}

func main() {
	utils.RunWithTimer(partOne) // main.partOne -- 239 -- took 0 ms
	utils.RunWithTimer(partTwo) // main.partTwo -- 141 -- took 1 ms
}

File: 2016/python/day1/main.py
from utils.timers import run_with_timer, get_data_with_timer


def get_data(filename):
	return [(x[0], int(x[1:])) for x in open(filename).readline().split(", ")]


DIRS = [(0, 1), (1, 0), (0, -1), (-1, 0)]


def get_points_between(start, end):
	pts = {end}.union({(i, start[1]) for i in range(start[0], end[0], -1 if start[0] > end[0] else 1)})
	return pts.union({(start[0], i) for i in range(start[1], end[1], -1 if start[1] > end[1] else 1)})


def part_one(d):
	curr_or = 0
	curr_pos = (0, 0)

	for x in d:
		curr_or = ((curr_or + 1) if x[0] == 'R' else (curr_or - 1)) % 4
		curr_pos = (curr_pos[0] + (DIRS[curr_or][0] * x[1]), curr_pos[1] + (DIRS[curr_or][1] * x[1]))
	return abs(curr_pos[0]) + abs(curr_pos[1])


def part_two(d):
	curr_or = 0
	curr_pos = (0, 0)
	visited = [curr_pos]

	for x in d:
		curr_or = ((curr_or + 1) if x[0] == 'R' else (curr_or - 1)) % 4
		next_pos = (curr_pos[0] + (DIRS[curr_or][0] * x[1]), curr_pos[1] + (DIRS[curr_or][1] * x[1]))
		line = get_points_between(curr_pos, next_pos)
		for y in line:
			if y in visited and y != curr_pos:
				return abs(y[0]) + abs(y[1])
			visited.append(y)
		curr_pos = next_pos
	return None


def main(f="input.txt"):
	data = get_data_with_timer(get_data, f)
	run_with_timer(part_one, data)
	run_with_timer(part_two, data)


if __name__ == '__main__':
	main()
